sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/ShardReadersPool.java

```diff
@@ -159,4 +212,91 @@ ShardRecordsIterator createShardIterator(SimplifiedKinesisClient kinesis,
     return new ShardRecordsIterator(checkpoint, kinesis);
   }
 
+  /**
+   * Waits until all records read from given shardRecordsIterator are taken from
+   * {@link #recordsQueue} and acked.
+   * Uses {@link #numberOfRecordsInAQueueByShard} map to track the amount of remaining events.
+   */
+  private void waitUntilAllShardRecordsRead(ShardRecordsIterator shardRecordsIterator)
+      throws InterruptedException {
+    while (!allShardRecordsRead(shardRecordsIterator)) {
+      Thread.sleep(TimeUnit.SECONDS.toMillis(1));
+    }
+  }
+
+  private boolean allShardRecordsRead(final ShardRecordsIterator shardRecordsIterator) {
+    return numberOfRecordsInAQueueByShard.get(shardRecordsIterator.getShardId()).get() == 0;
+  }
+
+  /**
+   * <p>
+   * Tries to find successors of a given shard and start reading them. Each closed shard can have
+   * 0, 1 or 2 successors
+   * <ul>
+   *   <li>0 successors - when shard was merged with another shard and this one is considered
+   *   adjacent by merge operation</li>
+   *   <li>1 successor - when shard was merged with another shard and this one is considered a
+   *   parent by merge operation</li>
+   *   <li>2 successors - when shard was split into two shards</li>
+   * </ul>
+   * </p>
+   * <p>
+   * Once shard successors are established, the transition to reading new shards can begin.
+   * The {@link ShardReadersPool#shardIteratorsMap}, {@link ShardReadersPool#recordsQueue}
+   * and {@link ShardReadersPool#numberOfRecordsInAQueueByShard} need to be updated atomically to
+   * remain in sync. Therefore the operation is synchronized on this {@link ShardReadersPool}
+   * instance. During this operation, the immutable {@link ShardReadersPool#shardIteratorsMap}
+   * is replaced with a new one holding references to {@link ShardRecordsIterator} instances for
+   * open shards only. The {@link ShardReadersPool#recordsQueue} is replaced with a new instance of
+   * capacity adjusted to the new number of open shards and records remaining in previous queue are
+   * drained to the new one. Also, the counter for already closed shard is removed from
+   * {@link ShardReadersPool#numberOfRecordsInAQueueByShard} map.
+   * </p>
+   * <p>
+   * Finally when atomic update is finished, new threads are spawned for reading the successive
+   * shards. The thread that handled reading from already closed shard can finally complete.
+   * </p>
+   */
+  private void readFromSuccessiveShards(final ShardRecordsIterator closedShardIterator)
+      throws TransientKinesisException {
+    List<ShardRecordsIterator> successiveShardRecordIterators = closedShardIterator
+        .findSuccessiveShardRecordIterators();
+
+    // Handle one closedShardIterator at a time to keep shardIteratorsMap,
```

Comment:
I'm really not a fan of having a mixture of synchronized blocks and atomic primitives :( Especially since the atomic primitives apparently need to be kept in sync with each other anyway. Can we use synchronized everywhere?

Reply:
I'm also not a fan but it seems to me like an acceptable trade-off. Changing this pool to use synchronized everywhere and do not use any atomics won't be easy. One problem I see is the blocking nature of recordsQueue - it can't be simply guarded with synchronized.

Reply:
I agree that recordsQueue should stay a BlockingQueue. However, are there any issues besides that? I think if we use `synchronized` for other synchronization, then you don't necessarily even need to swap out the queue. E.g. you could just create a queue with a particular maximum capacity, corresponding to maximum possible number of shards, which if I understand correctly is not that big (a few hundred).

Reply:
A queue with capacity corresponding to maximum possible number of shards could lead to running out of memory. When runner is calling reader's advance() method not fast enough it could quickly fill too much.